VirtoCommerce.Storefront/Domain/Cart/CartBuilder.cs

```diff
@@ -332,17 +332,11 @@ public virtual async Task MergeWithCartAsync(ShoppingCart cart)
         {
             EnsureCartExists();
 
-            //Reset primary keys for all aggregated entities before merge
-            //To prevent insertions same Ids for target cart
-            //exclude user because it might be the current one
-            var entities = cart.GetFlatObjectsListWithInterface<IEntity>();
-            foreach (var entity in entities.Where(x => !(x is User)).ToList())
-            {
-                entity.Id = null;
-            }
 
-            foreach (var lineItem in cart.Items)
+
```

Comment:
You need to reset ids for the cloned original cart instance.  Just add `cart.Clone()` before ids resetting.

Reply:
OK